libs/elasticsearch-core/src/main/java/org/elasticsearch/bootstrap/JarHell.java

```diff
@@ -0,0 +1,289 @@
+            // Technically empty classpath element behaves like CWD.
+            // So below is the "correct" code, however in practice with ES, this is usually just a misconfiguration,
+            // from old shell scripts left behind or something:
+            //   if (element.isEmpty()) {
+            //      element = System.getProperty("user.dir");
+            //   }
+            // Instead we just throw an exception, and keep it clean.
+            if (element.isEmpty()) {
+                throw new IllegalStateException("Classpath should not contain empty elements! (outdated shell script from a previous " +
+                    "version?) classpath='" + classPath + "'");
+            }
+            // we should be able to just Paths.get() each element, but unfortunately this is not the
+            // whole story on how classpath parsing works: if you want to know, start at sun.misc.Launcher,
+            // be sure to stop before you tear out your eyes. we just handle the "alternative" filename
+            // specification which java seems to allow, explicitly, right here...
+            if (element.startsWith("/") && "\\".equals(fileSeparator)) {
+                // "correct" the entry to become a normal entry
+                // change to correct file separators
+                element = element.replace("/", "\\");
+                // if there is a drive letter, nuke the leading separator
+                if (element.length() >= 3 && element.charAt(2) == ':') {
+                    element = element.substring(1);
+                }
+            }
+            // now just parse as ordinary file
+            try {
+                URL url = PathUtils.get(element).toUri().toURL();
+                if (urlElements.add(url) == false) {
+                    throw new IllegalStateException("jar hell!" + System.lineSeparator() +
+                        "duplicate jar [" + element + "] on classpath: " + classPath);
+                }
+            } catch (MalformedURLException e) {
+                // should not happen, as we use the filesystem API
+                throw new RuntimeException(e);
+            }
+        }
+        return Collections.unmodifiableSet(urlElements);
+    }
+
+    /**
+     * Checks the set of URLs for duplicate classes
+     * @throws IllegalStateException if jar hell was found
+     */
+    @SuppressForbidden(reason = "needs JarFile for speed, just reading entries")
+    public static void checkJarHell(Set<URL> urls) throws URISyntaxException, IOException {
+        Logger logger = Loggers.getLogger(JarHell.class);
+        // we don't try to be sneaky and use deprecated/internal/not portable stuff
+        // like sun.boot.class.path, and with jigsaw we don't yet have a way to get
+        // a "list" at all. So just exclude any elements underneath the java home
+        String javaHome = System.getProperty("java.home");
+        logger.debug("java.home: {}", javaHome);
```

Comment:
I wonder if we should push a `BiConsumer<String, String>` through here, then server can invoke this with `(format, parameter) -> logger.debug(format, parameter)` and we do not have to have the logging dependency brought in here? I would like to put off pulling in the logging dependency as long as possible, at least until we have a better idea how to do it without `log4j2-core` having to come along for the ride. The only thing this loses us is logger usage format checks; this might be okay but maybe @ywelsch has a suggestion on how we can apply the checks here too.

Reply:
I added a commit that removed the pulling in of `log4j2-core`. We only depend on api now.